server/repository/migrations/postgres/2022-06-07T15-00_create_document_registry/up.sql

```diff
@@ -0,0 +1,15 @@
+CREATE TYPE document_context AS ENUM (
+    'PATIENT',
+    'PROGRAM',
```

Comment:
Shouldn't programs have a seperate table ? It's a different 'thing' to to documents, and doesn't have parent_id or document_type or document_context or form_schema-id.

Reply:
In the program document patient specific data for a program is stored, e.g. family history, or if the patient is in a risc group. This information can in general change over time. Every patient can only have one program of each kind, i.e. you can't be enrolled in the same program twice.

Reply:
ok I dont' quite understand how this works (i did look at the issue).

What would data look like for this scenario:

two programs, 'HIV' and 'TB'.

`HIV` -> has document type 'Encounter A' (with two versions of ui schema and json schema)
`TB` -> has document type 'Encounter B' (just with one version)

There is a base 'Patient' document (not linked to any program, for patient data entry for stores with no programs), with two versions of ui schema and json schema

One common 'Patient' document to be used for both `TB` and `HIV`

By two versions btw I mean customer has added new fields to the form.

Reply:
There is no document_versions yet, just the config

Reply:
Is your question how to manage multiple schema versions? e.g. what happens if an older document version has a different schema?

The registry describes what schemas should be used for new document updates, e.g. when creating or updating a document.
If there is an existing document that already uses an old schema version, the old schema id is recorded in the document. 

The central json schemas should be immutable, i.e. when the a schema is update the registry entry needs to be updated to point to this new schema. For ui schema its probably fine/desirable to always show the latest UI as specified in the registry. 


A patient could have the following docs:

`patients/patient1` (type: Patient, context: Patient)
`patients/patient1/programs/HIVProgram` (type: HIVProgram, context: Program)
`patients/patient1/programs/HIVProgram/encounters/EncounterA_1` (type: EncounterA, context: Encounter)
`patients/patient1/programs/HIVProgram/encounters/EncounterA_2` (type: EncounterA, context: Encounter)
`patients/patient1/programs/TBProgram` (type: TBProgram, context: Program)
`patients/patient1/programs/TBProgram/encounters/EncounterB_1` (type; EncounterB, context: Encounter)

Reply:
Can you please show what document_registry and document_hierarchy rows look like for the example I gave, I can't figure it out (not including actual document data), just the document_hierarchy and document_registry rows.

Reply:
I think I might be missing something still, i would have thought that `program` is it's own type of record. i.e. two tables document_registry and program (as per the example I gave):

program.

| id | name |
|----|------|
| 1  | HIV  |
| 2  | TB   |

document_registry

| id | program_id (optional) | name        | version | form_schema_id | is_encounter (or is_tracker) |
|----|-----------------------|-------------|---------|----------------|------------------------------|
| 1  |                       | Patient     | 1       | 1              | false                        |
| 2  |                       | Patient     | 2       | 2              | false                        |
| 3  | 1                     | Patient     | 1       | 3              | false                        |
| 4  | 2                     | Patient     | 1       | 3              | false                        |
| 5  | 1                     | Encounter A | 1       | 4              | true                         |
| 6  | 1                     | Encounter A | 2       | 5              | true                         |
| 7  | 2                     | Encounter B | 1       | 6              | true                         |

permissions.

Permissions will need to be defined for user/store and program (can link to program in above schema).

I am have a meeting now, was going to add a little bit more, will do afterwards

Reply:
Some initial questions:

- why do you need a programs table? A program is just a special kind of document. If you want to add a program you can just add an entry to the document_registry table (programs have schemas as well).
- why version? when would you use this? or is it just for completeness?
- what do you mean by is_tracker?

Could you please give me your motivation behind these schemas than I can explain how it works with mine.

Reply:
Hmm, i see how your schema can work now, motivation for schema i outlined comes from what I thought the requirements were (based on what I understand about the functionality we are building, is it correct ? do we need Mark Glower/Richard to have a weigh in ?)

1. General Patient form should be available (with general patient information)
  a. General Patient form should be customisable (the latest version of the form should always be used)
  b. Newer version of General Patient form should be backwards compatible with previous version (because it's pre populated from previous version)
  c. On edit or create, some 'required' information from General Patient form should be save in patient record
2. A program can be configured in the system, and collection of patient forms can be associated with the program
  a. User access to a program is configured (by store?)
  b. There are two types of program forms, encounter or tracker (multiple entries for the same patient)
  c. And single entry (but can be edited multiple times)
  d. Previous entries for encounter or tracker form can be edited (in case of error)
  e. Tracker and single entry forms should be customisable (the latest version of the form should be used)
  f. Newer version of single entry forms should be backwards compatible with previous version (because it's pre populated from previous version)
  g. Newer version of tracker forms should be backwards compatible with previous version ? (for comparison with previous versions ?)
  h. Program single entry form can edit patient details in similar fashion to General Patient form
  i. New single entry form for program can be pre populated from patient details or General Patient form ?
3. Patient can be enrolled into a program
4. All edits to form entries are recorded in history

New encounter/tracker entry should be pre populated with some data ? From previous encounter ? From patient info ?

I've added question marks just to clarify, not sure about those. But the schema I was suggesting was not based on the question mark points.

I think the best way forward for us to reach consensus would be through requirements confirmed by stakeholder (or where we are in 100% agreements about them)

Reply:
Ohh man, i delete a comment I shouldn't have, i wrote one listing `requirements` then another one that started with `I reverted a comment` (and deleted both by accident), I am wondering if you can see the last one in your email ? Can you paste it in if it's there ?

Reply:
If you don't find the comment, I'll re-write it. I think atm I am leaning towards:

program

| id | name    |
|----|---------|
| 1  | General |
| 2  | HIV     |
| 3  | TB      |

document_type

| id | name        | version | form_schema_id | is_encounter (or is_tracker) |
|----|-------------|---------|----------------|------------------------------|
| 1  | Patient     | 1       | 1              | false                        |
| 2  | Patient     | 2       | 2              | false                        |
| 3  | Encounter A | 1       | 3              | true                         |
| 4  | Encounter A | 2       | 4              | true                         |
| 5  | Encounter B | 1       | 5              | true                         |

document_type_program_join

| id | program_id | document_type_id |
|----|------------|------------------|
| 1  | 1          | 1                |
| 2  | 1          | 2                |
| 3  | 2          | 2                |
| 4  | 2          | 3                |
| 5  | 2          | 4                |
| 6  | 3          | 1                |
| 7  | 3          | 6                |

Reply:
Accidentally deleted comment from Andrei:

> I reverted a comment, just in case you've seen it in email notifications.
> 
>     why do you need a programs table
> 
>     Program (HIV, TB, Malaria) is a broader functionality, and is used in other areas, including our speciality the supply chain (they are the 'same thing' in terms of reporting, and it's one area where HIS and supply chain data meet). So to me, they are not related to documents but as their own entity to which documents belong.
>     It doesn't sound like we have a requirement to link document types with other types in hierarchy, so it seems parent_id only use is to link encounters to programs (would encounters ever be linked to other encounters ? if so can you give example from requirements)
> 
>     A program is just a special kind of document
> 
>     As per above, in terms of full context, programs are beyond document.
> 
>     why version? when would you use this? or is it just for completeness?
> 
>     Since form and json schema can change for a document_type (it's customisable), for some reason I assumed the previous schema needs to be records. On second thought the only immediate use case I can see of this is for user to look at historic schemas to re-introduce something. So not sure if it's needed.
> 
>     what do you mean by is_tracker
> 
>     I thought there were two types of forms, one time with multiple edits (i.e. patient) and tracker which is HIS term for repeating data point, which is encounter in our case.
> 
>     Could you please give me your motivation behind these schemas than I can explain how it works with mine.
> 
> Basically the above. I realise my example is not great, especially One common 'Patient' document type to be used for both TB and HIV <- i think in country, we would always just have one Patient document type, and program related document are extras. i.e. when looking at program view in UI for a particular patient, base Patient document type is present in one form, any extra non tracker forms for the program and any extra tracker forms for the program.

Reply:
Thank you, i have something else to add, but I'll do it in the issue